### PR TITLE
TicTacToe demo cocos2dx functionality 

### PR DESCRIPTION
**Pull request by Grant-Postma**

### CODE REVIEW

File: database/testapp/src/common_main.cc [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
 #include <algorithm>
 #include <ctime>
+
```
Comment (alexames) [inline database/testapp/src/common_main.cc:17]: Since you're not really touching this file, can you just revert this change completely?

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#include <cstdlib>
+#include <iostream>
+
+#include "cocos2d.h"
+
+USING_NS_CC;
+
+const int kNumberOfTiles = 9;
+const int kMaxMovesPerPlayer = 1 + kNumberOfTiles / 2;
+const int kTilesX = std::sqrt(kNumberOfTiles);
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:13]: I would do this the other way:

kTilesX = 3
kTilesY = 3
kNumberOfTiles = kTilesX * kTilesY;

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <array>
+#include <cstdlib>
+#include <iostream>
+
+#include "cocos2d.h"
+
+USING_NS_CC;
+
+const int kNumberOfTiles = 9;
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:11]: Make all of these static.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+const int kMaxMovesPerPlayer = 1 + kNumberOfTiles / 2;
+const int kTilesX = std::sqrt(kNumberOfTiles);
+const int kTilesY = std::sqrt(kNumberOfTiles);
+const double kScreenWidth = 600;
+const double kScreenHeight = 600;
+const double kTileWidth = (kScreenWidth / kTilesX);
+const double kTileHeight = (kScreenHeight / kTilesY);
+const int kNumberOfPlayers = 2;
+std::array<std::string, kNumberOfPlayers> kMoveImageFileNames = {
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:20]: We can't have std::strings as global variables according to the C++ style guide. 

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

strings do dynamic allocation and then delete on destruction, which is against the style guide rules. 

Instead either use const char* here, or make it a string* and initialize the value at startup.

File: demos/TicTacToe/Classes/AppDelegate.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include "cocos2d.h"
+
+class AppDelegate : private cocos2d::Application {
+ public:
+  AppDelegate();
+  virtual ~AppDelegate();
```
Comment (alexames) [inline demos/TicTacToe/Classes/AppDelegate.h:8]: Instead of virtual, use override.

https://en.cppreference.com/w/cpp/language/override

File: demos/TicTacToe/Classes/TicTacToeScene.h [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#pragma once
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.h:1]: Instead of pragma once, use header guard:

https://google.github.io/styleguide/cppguide.html#The__define_Guard

File: demos/TicTacToe/Classes/TicTacToeScene.h [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+#include <array>
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.h:3]: I don't think you need any of these in the header. None of them are needed for what is exposed in this header.

File: demos/TicTacToe/Classes/TicTacToeScene.h [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+#include <cstdlib>
+#include <iostream>
+
+#include "cocos2d.h"
+
+class TicTacToe : public cocos2d::Layer {
+ public:
+  static cocos2d::Scene* createScene();
+  virtual bool init();
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.h:12]: override

File: demos/TicTacToe/Classes/TicTacToeScene.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+#include <array>
+#include <cstdlib>
+#include <iostream>
+
+#include "cocos2d.h"
+
+class TicTacToe : public cocos2d::Layer {
+ public:
+  static cocos2d::Scene* createScene();
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.h:11]: Is this a function that Cocos expects to exist, or did you create it? Unless it's a cocos library function, it should be PascalCase. 

https://google.github.io/styleguide/cppguide.html#Function_Names

File: demos/TicTacToe/Classes/AppDelegate.h [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#pragma once
```
Comment (alexames) [inline demos/TicTacToe/Classes/AppDelegate.h:1]: Instead of pragma once, use header guard:

https://google.github.io/styleguide/cppguide.html#The__define_Guard

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  auto file_names_it = std::begin(kMoveImageFileNames);
+
+  // TODO(grantpostma): This should reflect the size that is set in AppDelegate.
+  // Should modify kTileWidth and kTileHeight based on that size auto
+  // kScreenWidth = Director::getInstance()->getWinSize().width; auto
+  // kScreenHeight =  Director::getInstance()->getWinSize().height;
+
+  auto board_sprite = Sprite::create("tic_tac_toe_board.png");
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:44]: Put "tic_tac_toe_board.png" in a constant

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  board_sprite->setPosition(0, 0);
+  board_sprite->setAnchorPoint(Vec2(0.0, 0.0));
+
+  auto touchListener = EventListenerTouchOneByOne::create();
+  touchListener->onTouchBegan = [=](Touch* touch,
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:49]: https://google.github.io/styleguide/cppguide.html#Lambda_expressions

> Prefer explicit captures when the lambda will escape the current scope.

Only capture the variables you need in the lambda.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+                      kTilesX * floor(touch->getLocation().y / kTileHeight);
+
+      auto sprite = Sprite::create(*file_names_it);
+      // This calculates and sets the position of the sprite based on the
+      // move_tile and the constant screen variables.
+      sprite->setPosition((.5 + move_tile % kTilesX) * kTileWidth,
+                          (.5 + move_tile / kTilesY) * kTileHeight);
+      board_sprite->addChild(sprite);
+
+      file_names_it++;
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:66]: This feels like an odd use of iterators to me. Iterators are mostly (but not exclusively) for iterating over things in a loop. For a case like this, I would just have a "current_player_index" or something, and set it like so:

    current_player_index = (current_player_index + 1) % kNumberOfPlayers;

File: demos/TicTacToe/Classes/AppDelegate.cpp [hunk 14]
```diff
@@ -0,0 +1,31 @@
+#include "AppDelegate.h"
+
+#include "TicTacToeScene.h"
+
+USING_NS_CC;
+
+extern const float kFrameWidth = 600;
+extern const float kFrameHeight = 600;
```
Comment (alexames) [inline demos/TicTacToe/Classes/AppDelegate.cpp:8]: `extern` means the value is defined elsewhere. 

Think of it this way: When you declare a function, you're implicitly declaring it extern, and promising that you'll define it elsewhere. 

It's the same thing here, except the extern has to be explicit (C++ is inconsistent about this. It's confusing until you understand what's going on). You declare the variable here, and somewhere else you define what the value is.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 15]
```diff
@@ -0,0 +1,81 @@
+#include "TicTacToeScene.h"
+
+#include <array>
+#include <cstdlib>
+#include <iostream>
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:5]: I don't think you're using iostream in here.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+  auto touchListener = EventListenerTouchOneByOne::create();
+  touchListener->onTouchBegan = [board_sprite, current_player_index](
+                                    Touch* touch,
+                                    Event* event) mutable -> bool {
+    auto bounds = event->getCurrentTarget()->getBoundingBox();
+
+    if (bounds.containsPoint(touch->getLocation())) {
+      // This takes the touch location and calculates which tile number [0-8]
+      // corresponds to that location
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:56]: nit: add a period.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+static const int kTilesY = 3;
+static const int kNumberOfTiles = kTilesX * kTilesY;
+static const int kMaxMovesPerPlayer = 1 + kNumberOfTiles / 2;
+static const double kScreenWidth = 600;
+static const double kScreenHeight = 600;
+static const double kTileWidth = (kScreenWidth / kTilesX);
+static const double kTileHeight = (kScreenHeight / kTilesY);
+static const int kNumberOfPlayers = 2;
+static const char* kBoardImageFileName = "tic_tac_toe_board.png";
+std::array<const char*, kNumberOfPlayers> kMoveImageFileNames = {
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:20]: Can we use something other than MoveImageFileNames?  The move term is confusing to me -- at first I thought we were moving the files names.  How about kPlayerTokenFileNames or kPlayerSymbolFileNames?

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    "tic_tac_toe_x.png", "tic_tac_toe_o.png"};
+
+Scene* TicTacToe::createScene() {
+  auto scene = Scene::create();
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:24]: Since the return parameter is defined as Scene*, my preference would be define the 'scene' variable as Scene* here as well instead of auto.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+  // TODO(grantpostma): This should reflect the size that is set in AppDelegate.
+  // Should modify kTileWidth and kTileHeight based on that size auto
+  // kScreenWidth = Director::getInstance()->getWinSize().width; auto
+  // kScreenHeight =  Director::getInstance()->getWinSize().height;
+
+  auto board_sprite = Sprite::create(kBoardImageFileName);
+  board_sprite->setPosition(0, 0);
+  board_sprite->setAnchorPoint(Vec2(0.0, 0.0));
+
+  auto touchListener = EventListenerTouchOneByOne::create();
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:48]: Since this will serve as an example, please add a one line comment as to what this next section of code is meant to do.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 16 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:56]: I would remove "This takes" and say something like "Calculate the tile number [0-8] which corresponds to the touch location."

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  touchListener->onTouchBegan = [board_sprite, current_player_index](
+                                    Touch* touch,
+                                    Event* event) mutable -> bool {
+    auto bounds = event->getCurrentTarget()->getBoundingBox();
+
+    if (bounds.containsPoint(touch->getLocation())) {
+      // This takes the touch location and calculates which tile number [0-8]
+      // corresponds to that location
+      int move_tile = floor(touch->getLocation().x / kTileWidth) +
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:57]: selected_tile?

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+                      kTilesX * floor(touch->getLocation().y / kTileHeight);
+
+      auto sprite = Sprite::create(kMoveImageFileNames[current_player_index]);
+      // This calculates and sets the position of the sprite based on the
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:61]: "Calculate and set..."

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+      // move_tile and the constant screen variables.
+      sprite->setPosition((.5 + move_tile % kTilesX) * kTileWidth,
+                          (.5 + move_tile / kTilesY) * kTileHeight);
+      board_sprite->addChild(sprite);
+      current_player_index = (current_player_index + 1) % kNumberOfPlayers;
+    }
+    return true;
+  };
+
+  Director::getInstance()
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:71]: Can any of these return null pointers?

File: demos/TicTacToe/Classes/AppDelegate.cpp [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+const float kFrameWidth = 600;
+const float kFrameHeight = 600;
+
+AppDelegate::AppDelegate() {}
+
+AppDelegate::~AppDelegate() {}
+
+bool AppDelegate::applicationDidFinishLaunching() {
+  auto director = Director::getInstance();
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/AppDelegate.cpp:15]: could director return null?

File: demos/TicTacToe/Classes/AppDelegate.cpp [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+  auto glview = director->getOpenGLView();
+  if (!glview) {
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/AppDelegate.cpp:17]: if (glview == nullptr) {

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 18 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:24]: Could any of these return a null pointer?

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 21 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:61]: Null check on sprite.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  int current_player_index = 0;
+  auto file_names_it = std::begin(kMoveImageFileNames);
+
+  // TODO(grantpostma): This should reflect the size that is set in AppDelegate.
+  // Should modify kTileWidth and kTileHeight based on that size auto
+  // kScreenWidth = Director::getInstance()->getWinSize().width; auto
+  // kScreenHeight =  Director::getInstance()->getWinSize().height;
+
+  auto board_sprite = Sprite::create(kBoardImageFileName);
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeScene.cpp:44]: null check

File: demos/TicTacToe/Classes/TicTacToeScene.h (hunk 9 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.h:10]: Some comments around each of these functions, as trivial as they might be, would be nice.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 18 above)
Comment (Grant-Postma) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:24]: Both autos were swapped to their respective type.

The null pointer question is a good one and I am not sure. I'll look at some of the cocos documentation for createScene() and init() then loop around here

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 22 above)
Comment (Grant-Postma) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:71]: Director::getInstance() returns a static Director* which I don't believe can ever be a null pointer. The firebase sample https://github.com/FirebaseExtended/cocos2dx-cpp-sample/blob/master/database/Classes/FirebaseDatabaseScene.cpp#L25 
follows the same logic.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 19 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:63]: I would put the comments on line 64 65 (below) abobve the "auto touchListener =" statement.

File: demos/TicTacToe/Classes/TicTacToeScene.cpp (hunk 22 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.cpp:93]: cool cool.

File: demos/TicTacToe/Classes/TicTacToeScene.h (hunk 9 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/TicTacToeScene.h:12]: My issue is resolved (thanks for adding the comments) but I think you still need ot address Alex's..?

File: demos/TicTacToe/Classes/AppDelegate.cpp [hunk 26]
```diff
@@ -21,6 +21,9 @@ bool AppDelegate::applicationDidFinishLaunching() {
   }
 
   auto scene = TicTacToe::createScene();
+  if (scene == NULL) {
+    exit(true);
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/AppDelegate.cpp:25]: exiting the application is dangerous if it's not done in the main method as other allocations could have occurred (beyond your domian) that aren't cleaned up.  What if you return false here instead?

Also what about the code above?  Do the director and glview need to be released, destroyed, etc?

File: demos/TicTacToe/Classes/AppDelegate.cpp [hunk 27]
```diff
@@ -22,7 +22,9 @@ bool AppDelegate::applicationDidFinishLaunching() {
 
   auto scene = TicTacToe::createScene();
   if (scene == NULL) {
-    exit(true);
+    free(director);
+    free(glview);
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/AppDelegate.cpp:26]: free doesn't call C++ destructors.  You should call delete instead.

BUT... I'm not sure how you're supposed to release these things created by Cocoas2D.  It might not be enough to deallocate them - you might need to call a method on the class first. Like end() or destroy(), etc. You should check the exampes to see how they clean things up when the demo terminates.

File: demos/TicTacToe/Classes/AppDelegate.cpp (hunk 23 above)
Comment (DellaBitta) [reply demos/TicTacToe/Classes/AppDelegate.cpp:15]: Any change here?

File: demos/TicTacToe/Classes/TicTacToeScene.h (hunk 9 above)
Comment (Grant-Postma) [reply demos/TicTacToe/Classes/TicTacToeScene.h:12]: Thanks for the reminder. The createScene() is a cocos library function.
